File: GOF/Builder/cpp/builder/StandardMazeBuilder.cpp
#include "StandardMazeBuilder.h"
#include "../maze/Maze.h"
#include "../maze/Room.h"
#include "../maze/Wall.h"
#include "../maze/Door.h"

using std::make_shared;

void StandardMazeBuilder::buildMaze()
{
	m_currentMaze = make_shared<Maze>();
}

void StandardMazeBuilder::buildRoom(int n)
{
	if (m_currentMaze->roomNo(n).lock())
	{
		std::shared_ptr<Room> room = make_shared<Room>(n);
		m_currentMaze->addRoom(room);
		room->setSide(Room::Direction::North, make_shared<Wall>());
		room->setSide(Room::Direction::South, make_shared<Wall>());
		room->setSide(Room::Direction::East, make_shared<Wall>());
		room->setSide(Room::Direction::West, make_shared<Wall>());
	}
}

void StandardMazeBuilder::buildDoor(int n1, int n2)
{
	std::weak_ptr<Room> r1 = m_currentMaze->roomNo(n1);
	std::weak_ptr<Room> r2 = m_currentMaze->roomNo(n2);
	std::weak_ptr<Door> door = make_shared<Door>(r1, r2);

	//TODO
	/*r1->setSide(commonWall(r1, r2), d);
	r2->setSide(commonWall(r1, r2), d);*/
}

std::shared_ptr<Maze> StandardMazeBuilder::getMaze()
{
	return m_currentMaze;
}

File: GOF/Builder/java/src/Builder/StandardMazeBuilder.java
package Builder;

import Maze.Maze;
import Maze.Room;
import Maze.Wall;
import Maze.Door;

public class StandardMazeBuilder extends MazeBuilder {

    private Maze m_currentMaze;

    @Override
    public void buildMaze() {
        m_currentMaze = new Maze();
    }

    @Override
    public void buildRoom(int n) {
        if(m_currentMaze.roomNo(n) == null) {
            Room room = new Room(n);
            m_currentMaze.addRoom(room);

            room.setSide(Room.Direction.North, new Wall());
            room.setSide(Room.Direction.South, new Wall());
            room.setSide(Room.Direction.East, new Wall());
            room.setSide(Room.Direction.West, new Wall());
        }
    }

    @Override
    public void buildDoor(int n1, int n2) {
        Room r1 = m_currentMaze.roomNo(n1);
        Room r2 = m_currentMaze.roomNo(n2);
        Door door = new Door();

        //TODO
        /*r1->setSide(commonWall(r1, r2), d);
        r2->setSide(commonWall(r1, r2), d)*/
    }

    @Override
    public Maze getMaze() {
        return m_currentMaze;
    }
}
